### ISSUE TITLE
Implementing MaxCUT

### ISSUE THREAD

**Issue by pboes** [dpsanders/SatisfiabilityInterface.jl#24]
Hey, I am trying to implement the decision version of simple MaxCUT to test some custom SAT solvers. I've tried various formulations that only use elements I already see in the examples, but somehow keep on getting an error. 

```julia
using SatisfiabilityInterface, Symbolics

## Formulation 1
function maxcut_problem(V, E, k)
  n = length(V)
  @variables c[1:n]
  constraints = [
	  [c[i] ∈ (0:1) for i in 1:n]
	  sum(c[i] != c[j] for (i,j) in E) <= k
  ]
  return DiscreteCSP(constraints)
end


## Formulation 2
function maxcut_problem_alt(V, E, k)
  n = length(V)
  ne = length(E)
  @variables c[1:n]
  @variables e[1:ne]
  constraints = [
	  [c[i] ∈ (0:1) for i in 1:n]
	  [e[i] ∈ (0:1) for i in 1:ne]
	  [e[i] == (c[E[i][1]] != c[E[i][2]]) for i in 1:ne]
	  sum(e[i] for i in 1:ne) <= k
  ]
  return DiscreteCSP(constraints)
end

V=[1,2]
E = [(1,2)]
maxcut_problem_alt(V,E, 1)
maxcut_problem(V,E,1)
```
Both give rise to essentially the same error:
```julia
MethodError: no method matching iterate(::SatisfiabilityInterface.DiscreteVariable{SymbolicUtils.Term{Real, Nothing}, Vector{Int64}, Vector{SymbolicUtils.Sym{Real, Nothing}}, Dict{Int64, SymbolicUtils.Sym{Real, Nothing}}})

Closest candidates are:

iterate(!Matched::Union{LinRange, StepRangeLen}) at range.jl:664

iterate(!Matched::Union{LinRange, StepRangeLen}, !Matched::Int64) at range.jl:664

iterate(!Matched::T) where T<:Union{Base.KeySet{var"#s77", var"#s76"} where {var"#s77", var"#s76"<:Dict}, Base.ValueIterator{var"#s75"} where var"#s75"<:Dict} at dict.jl:693

...

    indexed_iterate(::SatisfiabilityInterface.DiscreteVariable{SymbolicUtils.Term{Real, Nothing}, Vector{Int64}, Vector{SymbolicUtils.Sym{Real, Nothing}}, Dict{Int64, SymbolicUtils.Sym{Real, Nothing}}}, ::Int64)@tuple.jl:89
    SatisfiabilityInterface.DiscreteCSP(::SatisfiabilityInterface.ConstraintSatisfactionProblem)@relations.jl:388
    SatisfiabilityInterface.DiscreteCSP(::Vector{Symbolics.Num})@relations.jl:368
    maxcut_problem_alt(::Vector{UnitRange{Int64}}, ::Vector{Tuple{Int64, Int64}}, ::Int64)@[Other: 14](http://localhost:1234/edit?id=5c6c88ce-fb87-11ec-195f-190fab30622d#)
    top-level scope@[Local: 4](http://localhost:1234/edit?id=5c6c88ce-fb87-11ec-195f-190fab30622d#)[inlined]
```

I am not sure whether what is going here is that the problem simply cannot be modeled using the supported class of CSPs in SatisfiabilityInterface.jl or whether I'm simply formulating the problem badly at the Symbolics.jl level. Any help appreciated.

**Comment by dpsanders** [comment]
Hi,

Thanks for your interest and apologies for the delay in replying. 

I believe the (or at least one) issue is that the new version of `@variables` in Symbolics.jl does not do what we need.
You can use the following workaround:

```jl
function make_vector(name::Symbol, range)
    return variable.(name, range)
end

b = make_vector(:b, 1:4)   # equivalent of old @variables b[1:4]
```

I think there should be a way to do this using `Symbolics.scalarize`.  cc @shashi

It would be great to have a macro to do this, e.g. `@vars b[1:4]`, which should be easy enough.

**Comment by pboes** [comment]
Thanks @dpsanders, 

this helped. After some fiddling the following does not complain and (so far) produces correct results

```julia
using Symbolics
using SatisfiabilityInterface

function var_vector(name::Symbol, range)
    return Num.(Variable.(name, range))
end

function maxcut_problem(V, E, k)
    n = length(V)
    ne = length(E)
    c = var_vector(:c, 1:n)
    e = var_vector(:e, 1:ne)
    constraints = [
        [c[i] ∈ (0:1) for i in 1:n]
        [e[i] ∈ (0:1) for i in 1:ne]
        [e[i] == (c[E[i][1]] - c[E[i][2]])^2 for i in 1:ne]
        sum(e[i] for i in 1:ne) == k
    ]
    return DiscreteCSP(constraints)
end

V = (1:4)
E = [(1, 2), (2, 3), (3, 4), (4, 1)]
maxcut = maxcut_problem(V, E, 4)
solve(maxcut)
```
Apart from scalarizing the variables as you suggested, I also had to fiddle with the formulation of third constraint. For example, formulating that constraint instead as 

```julia
 [c[E[i][1]] != c[E[i][2]] for i in 1:ne if isequal(e[i],1)]
```
did generate a DiscreteCSP without complaining but upon inspecting false solutions turns out the if clause was basically ignored.

Happy to PR this as an additional example and turn the var_vector into a macro, if interested :)

BTW I'm really curious what method you use to translate CSPs into SAT instances :)

**Comment by dpsanders** [comment]
That's a very nice formulation! Basically you were struggling with the fact that there was no way to check if `v[i]` and `v[j]` were different.

I have just merged a refactor where I added unary functions `abs` and `sign`.
There is also `delta`, which satisfies that `delta(x)` is 1 if `x == 0`, and 0 otherwise.

I will be adding some more functionality that should make this kind of thing easier, and also will be adding `maximize` soon.

Also I intend to add some docs about the encoding. It's the simplest unary encoding, where a discrete variable `x` with `n` values is modelled by `n` Boolean variables, each of which says whether `x == i` or not.
